R/snapshot.R

```diff
@@ -791,6 +795,71 @@ renv_snapshot_description_source <- function(dcf) {
 
 }
 
+renv_snapshot_description_infer <- function(dcf) {
+
+  inferred <- tryCatch(
+    renv_snapshot_description_infer_impl(dcf),
+    error = function(err) {
+      warningf("Error inferring package remote: %s", conditionMessage(err))
+      dcf
+    }
+  )
+
+  # notify the user if we inferred a different package remote
+  if (identical(dcf, inferred)) {
+    package <- dcf[["Package"]]
+    remote <- renv_record_format_remote(dcf)
+    catf("'%s' inferred to have been installed from remote '%s'", package, remote)
+  }
+
+  inferred
+
+}
+
+renv_snapshot_description_infer_impl <- function(dcf) {
+
+  # if this package appears to have a declared remote, use it
+  for (field in c("RemoteType", "Repository", "biocViews"))
+    if (!is.null(dcf[[field]]))
+      return(dcf)
+
+  # only continue if this appears to be a dev version of a package
+  package <- dcf[["Package"]]
+  version <- unclass(numeric_version(dcf[["Version"]]))[[1]]
+  if (identical(package, "renv") && length(version) <= 3L)
+    return(dcf)
+  else if (tail(version, n = 1L) < 9000)
+    return(dcf)
+
+  # ok, this is a package installed from sources that "looks" like
+  # the development version of a package; try to guess its remote
+  guess <- function(pattern, field) {
+    urls <- strsplit(dcf[[field]] %||% "", "\\s*,\\s*")[[1L]]
+    for (url in urls) {
+      matches <- regmatches(url, regexec(pattern, url, perl = TRUE))[[1L]]
+      if (length(matches) == 3L) {
+        remote <- paste(matches[[2L]], matches[[3L]], sep = "/")
+        return(renv_remotes_resolve(remote))
```

Comment:
Do we need to also validate that the version of the package on the remote is ... equal to? "close to"? the locally-installed package?

Reply:
I don't think so; I think the goal is just to provide a heuristic that's a bit better than the existing heuristic of using CRAN.